FactoryTools.lib.nut

```diff
@@ -0,0 +1,152 @@
+/**
+* Copyright (C) 2015-2018 Electric Imp, Inc
+* Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files
+* (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge,
+* publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do
+* so, subject to the following conditions:
+* The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+* FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+* WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/**
+* FACTORY TOOLS LIBRARY utility for simplifying factory firmware flow.
+* @author cat-haines
+* @author ersatzavian
+* @author Tony Smith <tony@electricimp.com>
+* @author Elizabeth Rhodes <betsy@electricimp.com>
+*
+* version 2.2.0
+*/
+
+class FactoryTools {
+
+    static VERSION = "2.2.0";
+
+    /**
+    * Checks imp.configparams table for factory firmware flag
+    * Params:
+    *   callback (optional) : Function with one param - boolean flag if imp is running factory firmware.
+    *                         On device called after imp has a chance populate imp.configparams table.
+    *                         On Agent called immediately.
+    * Returns:
+    *   bool (if no callback provided) or null (if callback is provided - bool is passed to callback)
+    */
+    static function isFactoryFirmware(callback = null) {
+        local isFF = ("factoryfirmware" in imp.configparams && imp.configparams["factoryfirmware"]);
+        if (callback) {
+            if (_isAgent()) {
+                callback(isFF);
+            } else {
+                imp.onidle(function() {
```

Comment:
Why is that (executing the callback from the onidle state)?
Why do we need to call `imp.onidle(null);`?

Reply:
To clear the on.idle() callback so that the callback isn't triggered again if the imp goes idle later.
on.idle() was used in earlier versions 2.1.1 and under because we can't be certain thatimp.configparams is populated *until* the imp goes idle, which is why pulling the configparams data into the variable *isFF* probably won't work. That change is not mine.

Reply:
My bad I should - I can't cache the results. Although I don't understand why we need to clear the onidle - that function will only ever trigger once, so if I'm in onidle then there is no need to clear it - that is redundant.

Reply:
that's my understanding too...